### DISCUSSION TITLE
Suggestions for future iterations of the USWDS mobile menu

### DISCUSSION THREAD

**Discussion by mariamarrero** [uswds/uswds#6626] [Proposals]
## Challenges with our old menu/navigation and reasons for updates

The team at USAGov made major enhancements	to the mobile menu of the website usa.gov. 

A few years ago, we adopted the USWDS code for the mobile menu. However, the information architecture of our website is broad (up to 5 levels deep) and users were not able to access all topics via the menu. We also noticed that the width of the old menu did not work very well with some of the long labels of our topics. We needed more horizontal space to be able to fit our content better. We wanted the menu to be more robust for both of our websites. 

## Research and exploration for a new menu concept
Data tells us that users come to usa.gov to find information and navigate within a page, then related topics within a section or category and finally they would navigate to a totally different section. Therefore, the need to display the difference topics and levels in the menu became critical for our site.

A few months ago we started working on a concept that would allow users to access any topic within our taxonomy from the menu in a simple and non cluttered way. 

After various sessions of brainstorming, testing and development, we implemented a mobile navigation with the following features/functionality:

- Users can navigate up to 5 levels deep in our IA without leaving the menu. 
- The menu has the ability to remember and display the current location/page (text in bold and red)
- The size of the menu (when opened/expanded) extends 100% in height and weight. This helps with long labels and the extra space reduces the need for wrapping text.
- The fact that the menu covers the entire screen also helps less web savvy users to focus on the content in the menu without having any background text behind it, which could be confusing to them.
- The word “Close” next to the “X” in the menu is spelled for accessibly purposes and it is a clear cue that the menu closes when tapped.
- The mobile menu is part of a sticky header, so it is present at all times even when users scroll down.
- The menu is 508 compliant and accessible.
- The menu has been tested with users for its usability and functionality.

During the course of this project we conducted user testing on working prototypes and finally with a high fidelity and working version of the menu. The results confirmed that this new mobile menu was helpful and better than the previous version. Users were successfully able to complete tasks that required navigation between topics while using the menu.

## Resources and additional info

Our team would be glad to share any additional information if requested. Attached are screenshots of the old and new versions of our mobile menus for comparison. You can also go to usa.gov (in a mobile device) and check it out in action.

**Old Menu - 2 levels deep**

![old-menu-2levels-deep](https://user-images.githubusercontent.com/6410374/63624077-4f3ce500-c5c9-11e9-9e75-221e85879071.jpg)

**New Menu - 5 levels deep**

![new-menu-5levels-deep](https://user-images.githubusercontent.com/6410374/63624084-54019900-c5c9-11e9-88e0-c57c0832ec8d.jpg)

Thanks for the opportunity to share!
Maria Marrero

**Comment by governmentSponsored** [comment]
@mariamarrero this is awesome feedback! We would love to see something like your `New Menu` is in the design system.

## Background on Commerce.gov and USWDS

The Commerce team is using something very similar to what you documented in that `Old Menu` structure on www.Commerce.gov. Commerce.gov utilizes the [USWDS Version 1 Drupal theme](https://www.drupal.org/project/uswds), which has one key difference--the ability to expand past level 2 menus (See screenshot below). It is not a perfect solution, but at least provides some good functionality to the end user.

In doing research and discovery to upgrade to USWDS Verison 2 Drupal theme, it was discovered that this key piece of functionality was removed to be more in line with the design system. Perfectly reasonable approach, but something that is a blocker going forward. There needs to be a way to provide users with a way to navigate to Level 3 and below when necessary. 

## Questions:
1. Could the usability research findings for the `New Menu` approach be shared?

## Suggestions
1. We agree that the `New Menu` approach is the direction this project should go. Usability and 508 compliance built in is absolutely key. 
1. For our team to move forward on USWDS v2, a bridge solution would have to be put in place. Neither are really *good* ideas, but rather would be *necessary* things to do in order to use USWDS v2. Our current position is to put USWDS v2 migration on hold until mobile menu is sorted out.
    * Keep USWDS v2 mobile menu. Add Level 3 and below menus to an accordion menu above the body content. Again, not ideal, but direction we would likely go to move forward with USWDS v2.
    * Refactor the USWDS v1 mobile menu to make it work in v2. This would require a significant amount of development work on our team that may need to be thrown out later on.

## Screenshots

### Current Commerce.gov on USWDS v1
![Image Pasted at 2019-9-19 16-05](https://user-images.githubusercontent.com/8822118/65278016-399fda00-daf9-11e9-80eb-59f5fc3bd60d.png)

### Local instance of Commerce.gov on USWDS v2
![Image Pasted at 2019-9-19 16-02](https://user-images.githubusercontent.com/8822118/65278029-402e5180-daf9-11e9-9906-90001d3d1d2a.png)

### Proposed Commerce.gov on USWDS v2 with menu above Body

Apologies for poor quality here, just some mock ups we have been throwing around.

![Proposed Mobile Menu Alternative Collapsed Body Menu](https://user-images.githubusercontent.com/8822118/65278079-563c1200-daf9-11e9-944c-05ef4041a546.png)

![Proposed Mobile Menu Alternative Expanded Body Menu](https://user-images.githubusercontent.com/8822118/65278084-589e6c00-daf9-11e9-85f5-418d0b51762c.png)

**Comment by mariamarrero** [comment]
Thanks @governmentSponsored! Yes, we hope so too!

So, one thing I would share with you is that we implemented a similar workaround or patch to our old menu as you did. We added a section at the top of mobile pages with links to levels 3, 4, and 5 of our information architecture. However, data showed us that users barely clicked on those links and we observed in usability tests that users would often miss it or skip it. They expected to find related links under the "Menu" and that is what lead us to explore a more robust navigation. 

The attached mockup was the temporary fix we implemented before our complete mobile navigation we have now. Hope this helps! 
![image (1)](https://user-images.githubusercontent.com/6410374/65434968-b2e74780-dded-11e9-8c4f-20c4f1ec6464.png)

**Comment by governmentSponsored** [comment]
@mariamarrero that is very helpful user experience feedback! We were worried folks might miss it or get confused about what menu items would be in which section (mobile menu or above body content). 

I think we'll hold off on adding that menu section above the body content, and give USWDS some time to sort out improvements to the mobile menu.
